File: 647-回文子串.cpp
class Solution {
public:
    int countSubstrings(string s) {
        int result = 0;
        int n = s.size();
        for(int mid = 0; mid <= 2 * n - 2; mid++){
            int left = mid / 2;
            int right = (mid + 1) / 2;
            while((left >= 0) && (right < n) && (s[left] == s[right])){
                left--;
                right++;
                result++;
            }
        }
        
        return result;
    }
};

File: 647-回文子串.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        index_set = set([(i, i) for i in range(len(s))])
        for left in range(len(s) - 1, -1, -1):
            for right in range(left, len(s)):
                if s[right] == s[left]:
                    if left + 1 > right - 1 or (left + 1, right - 1) in index_set:
                        index_set.add((left, right))

        return len(index_set)
